### DISCUSSION TITLE
FR: Browser Field not firing select event / endpoint

### DISCUSSION THREAD

**Discussion by madsem** [area17/twill#1001] [Q&A]
The Browser Field currently does not fire the endpoint when an item is selected/attached. It should be added though, to make it easier to use the browser with external data that might be downloaded from an API, on select of an item.

**EDIT**
yeah sorry should have explained better.
Currently it fires only on open and on search (enter key), not when you click the button to select one.

But this could be valuable because imagine you retrieve data from an external source like an API.

Do you want to download images & data for ALL results and write it to a table, or just the one the user selected?

If you don’t know which one the user selected, you have to write everything in the database before you can deliver suggestions for the browser to display.
(in my case this can be two dozen items plus 2-20 HQ screenshots per item, while image download could be queued, the items not as you obviously need the ID in the db first)

It could maybe attach different url parameter values depending on state...

onopen = &status=opened
onsearch = &status=search
onselect = &status=selected

**Comment by ifox** [comment]
Hey @madsem, what do you mean when it is selected/attached? Clicking the checkbox?
The endpoint is triggered each time you open the browser again, no?

**Comment by madsem** [comment]
@ifox okay, I totally forgot about the `ModuleRepository` methods for `afterSave`, `prepareFieldsBeforeCreate` etc.

That actually solves this issue, because I can load the suggestions into the `Browserfield` without saving them all to DB first, and then update the selected suggestion in the `afterSave` method.

Or is another one of those methods better suited for this? Just asking as I'm a bit foggy on which method is executed when during runtime.

I'll close this as I consider it a non-issue now.

**Comment by madsem** [comment]
Actually going to reopen this, because I just realized that even with the `afterSave` etc. methods, it is not possible to really integrate external data in a way that works smoothly.

The problem I am facing now is that, there is also additional data (images). So I can open the browser, make a search and get results returned from the Controller.

These results then don't have the primary key (ID) yet, that is needed to save the relationship to other crud modules. This can be handled easily in the `afterSave` method.

But, what about images / media files etc?

Those can be downloaded in the background, yes. But then they are only available for selection in a `medias` type field, AFTER the entity was saved.

Suppose you have a form like this:

Title
api browser
hero image (medias)

Now you select the content from the api browser, and in the next step you'd like to select an image from the selection you made in the api browser... Not possible without saving the entry first.

It would be possible though, if the `endPoint` was also triggered on clicking the "Attach ..." button.

Does that make any sense? Or is there a better solution to this that I am forgetting?

**EDIT**
PS: I've also thought about making a Crud module for the external content... But that brings new challenges with it, as a `Title` is always required and pops open in the modal when creating a new entry. But the `Title` would be redundant, as it comes via external API, so the editors would always have to create a title, which is then overwritten. Also external content would need to be created before related content in Twill crud modules could reference this content through `browser` fields... So there are drawbacks to this method as well.
